pgml-docs/docs/gym/fraud.md

```diff
@@ -0,0 +1,460 @@
+
+We can construct a query that provides a summary view of our orders, including the fraudulent label:
+
+```sql
+CREATE VIEW orders_summaries AS
+SELECT 
+  orders.id AS order_id, 
+  orders.customer_name,
+  payments.amount AS total, 
+  ARRAY_AGG(products.name) AS product_names,
+  CASE WHEN chargebacks.id IS NOT NULL 
+    THEN true 
+    ELSE false 
+  END AS fraudulent
+FROM orders
+LEFT JOIN payments ON payments.order_id = orders.id
+LEFT JOIN chargebacks ON chargebacks.payment_id = payments.id
+LEFT JOIN line_items ON line_items.order_id = orders.id
+LEFT JOIN products ON products.name = line_items.product_name
+GROUP BY 1, 2, 3, 5
+ORDER BY orders.id;
+```  
+
+```sql
+SELECT * FROM orders_summaries;
+```
+
+It's intuitive that thieves will be more attracted to gold bars, than a head of lettuce because the resell value is better. Perishable goods are more difficult to move on the black market. A good feature for our model would be the percentage of the order that is perishable. We can construct a query to return this feature for each order, along with the chargeback label. 
+
+```sql
+CREATE VIEW fraud_samples AS
+SELECT 
+  SUM(CASE WHEN products.perishable THEN (count * price) ELSE '$0.0' END) / SUM(payments.amount) AS perishable_percentage, 
+  CASE WHEN chargebacks.id IS NOT NULL 
+    THEN true 
+    ELSE false 
+  END AS fraudulent
+FROM orders
+LEFT JOIN payments ON payments.order_id = orders.id
+LEFT JOIN chargebacks ON chargebacks.payment_id = payments.id
+LEFT JOIN line_items ON line_items.order_id = orders.id
+LEFT JOIN products ON products.name = line_items.product_name
+GROUP BY orders.id, chargebacks.id
+ORDER BY orders.id;
+```
+
+```sql
+SELECT * FROM fraud_samples;
+```
+
+Training a model
+----------------
+
+This is a great training set for a machine learning model. We've found a feature that perfectly correlates with the label. This will allow models to generalize the fact based on our experience. Perishable orders are less likely to result in a chargeback. Now that we have a `VIEW` of this data, we can train a "classification" model to classify the features as fraud or not.
+
+```sql
+SELECT * FROM pgml.train(
+  project_name => 'Our Fraud Model', -- a friendly name we'll use to identify this machine learning project
+  task => 'classification', -- we want to classify into true or false
+  relation_name => 'fraud_samples', -- our view of the data
+  y_column_name => 'fraudulent' -- the "labels"
+);
+```
+
+Oops. We're going to get an error:
+
+```
+ERROR:  ValueError: This solver needs samples of at least 2 classes in the data, but the data contains only one class: False
```

Comment:
![image](https://user-images.githubusercontent.com/9561483/184193010-8d6935d0-05ab-4e08-8fa4-865dedc8d161.png)

Got that instead.

Reply:
![image](https://user-images.githubusercontent.com/9561483/184193406-cfe84134-a019-4c16-879b-da8f90aba83d.png)